enhancements/ingress/mutable-publishing-scope.md

```diff
@@ -0,0 +1,348 @@
+endpoint publishing strategy type.
+
+## Motivation
+
+Users sometimes wish to change the scope of an IngressController after it has
+been created.  Without this enhancement, it is possible to delete and recreate
+an IngressController to change its scope.  This enhancement makes it possible to
+change the scope of an IngressController _in situ_ on cloud platforms that
+support doing so, without deleting and recreating it.  This capability includes
+the default IngressController as well as custom IngressControllers.
+
+### Goals
+
+1. The cluster administrator can change the scope of the service load-balancer of an IngressController with "LoadBalancerService" endpoint publishing strategy type.
+2. If the cloud platform supports mutating scope on a service load-balancer, the operator will perform this operation when the cluster administrator changes the configured scope.
+3. If the cloud platform does not support mutating scope, the operator will advise the user how to complete the operation.
+
+### Non-Goals
+
+1. The operator does not delete a service load-balancer for an IngressController that is not marked for deletion.
+
+## Proposal
+
+Some platforms (such as Azure and GCP) support changing the scope of a service
+load-balancer between internal and external without deleting and recreating the
+load balancer, by setting an annotation on the Kubernetes Service object.  On
+these platforms, the operator merely sets the annotation to the desired scope,
+and the operation of changing it scope is complete.
+
+Other platforms (such as AWS) require deleting and recreating a load balancer to
+change its scope.  This operation is disruptive: It interrupts ingress traffic
+and may cause the load balancer's address to change.  On these platforms, the
+operator signals that the user must delete the Kubernetes Service object.  Once
+the user performs this step, the operator recreates the load balancer with the
+desired scope to complete the operation.
+
+In order to signal that the user must take action, the operator sets the
+`Progressing` status condition to `True` with a message that indicates the
+action that the user must take:
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: custom
+  namespace: openshift-ingress-operator
+spec:
+  replicas: 2
+  domain: mycluster.com
+  endpointPublishingStrategy:
+    type: LoadBalancerService
+    loadBalancer:
+      scope: Internal
+status:
+  availableReplicas: 2
+  conditions:
+  - message: Deployment has minimum availability.
+    reason: MinimumReplicasAvailable
+    status: "True"
+    type: Available
+  - message: Have load balancer with scope "External", want load balancer with scope "Internal".  You can delete the openshift-ingress/router-default service to proceed [...].  Alternatively, you can change the IngressController's spec.endpointPublishingStrategy.loadBalancer.scope field value back to its previous value [...].
```

Comment:
The message is really long and probably too much information.  Could we get away with something like: "Scope changed from ___ to ___, requires delete of service ___"

Reply:
The detailed wording is the result of some discussion here:  https://github.com/openshift/cluster-ingress-operator/pull/582#discussion_r605151494

The explicit and detailed message is intended to prevent mistakes and reduce support cases.  I can try to make this more concise, but I think it is important to communicate the options and their consequences because otherwise users will break their clusters' ingress.